src/liballoc/sync.rs

```diff
@@ -2150,3 +2269,12 @@ impl<T: ?Sized> AsRef<T> for Arc<T> {
 
 #[stable(feature = "pin", since = "1.33.0")]
 impl<T: ?Sized> Unpin for Arc<T> { }
+
+/// Computes the offset of the data field within ArcInner.
+unsafe fn data_offset<T: ?Sized>(ptr: *const T) -> isize {
+    // Align the unsized value to the end of the ArcInner.
+    // Because it is ?Sized, it will always be the last field in memory.
+    let align = align_of_val(&*ptr);
+    let layout = Layout::new::<ArcInner<()>>();
+    (layout.size() + layout.padding_needed_for(align)) as isize
```

Comment:
This seems very similar to some of the offset logic around `Arc::from/to_raw`. Can code be shared here?

Reply:
It actually is that exact code, extracted to a function and called from both places. See the deletions in `Arc::from_raw`.

Reply:
Ah, I had missed that. Awesome :)